src/backend/distributed/test/connection_cache.c

```diff
@@ -129,22 +130,30 @@ get_and_purge_connection(PG_FUNCTION_ARGS)
 
 /*
  * set_connection_status_bad does not remove the given connection from the connection hash.
- * It only sets its status to CONNECTION_BAD. On success, it returns true.
+ * It simply shuts down the underlying socket. On success, it returns true.
  */
 Datum
 set_connection_status_bad(PG_FUNCTION_ARGS)
 {
 	char *nodeName = PG_GETARG_CSTRING(0);
 	int32 nodePort = PG_GETARG_INT32(1);
+	int socket = -1;
+	int pqStatus PG_USED_FOR_ASSERTS_ONLY = 0;
 
 	PGconn *connection = GetOrEstablishConnection(nodeName, nodePort);
 	if (connection == NULL)
 	{
 		PG_RETURN_BOOL(false);
 	}
 
-	/* set the connection status */
-	SetConnectionStatus(connection, CONNECTION_BAD);
+	/* Prevent further reads/writes... */
+	socket = PQsocket(connection);
+	shutdown(socket, SHUT_RDWR);
```

Comment:
I also had a hunch to check the return value of `shutdown`, so I've addressed that. For the second bit, we can wait to refactor into a separate function until we have a need for it. I only moved it to a separate file in the first place because I wanted it to have a distinct set of `#include`s.